### PR TITLE
PR: Add a new Application plugin

### PR DESCRIPTION
**Pull request by dalthviz**
TODO:

- [x] Move application level actions from the Mainmenu and Help plugins to a new Application plugin
- [x] Move application level actions from the mainwindow to the Application plugin:
   - [x] Windows env variables action
   - [x] ~PYTHONPATH Manager~ (PYTHONPATH will be its own plugin)
   - [x] Restart
   - [x] ~Quit~ (part of the Console plugin)
- [x] Add actions to the main toolbar using the Toolbar plugin
     - [x]  Fix Toolbar plugin before param. Add `before_section` param for items addition.
- [x] Move general preferences to the Application plugin.
     - [x] Language
     - [x] Rendering engine
     - [x] Use single instance
     - [x] Prompt when exiting
     - [x] Show Internal Spyder errors to report them in Github
     - [x] Check for updates
     - [x] Screen resolution (high DPI screens)
          - [x] Normal
          - [x] Enable auto high DPI scaling
          - [x] Set custom high DPI scaling
     - [x] ~Vertical title bars in panes~ (Removed)
     - [x] Vertical tabs
     - [x] ~Animated toolbars and panes~ (Removed)
     - [x] ~Tear off menus~ (Removed)
     - [x] Custom margin for panes
     - [x] Cursor bliking
- [x] Fix possible failing tests
- [x] Dependencies

Part of #13527

### CODE REVIEW

**Review by steff456 — APPROVED** [review]
This PR is ready to merge! Thanks @dalthviz :)

File: spyder/plugins/application/container.py [hunk 1]
```diff
@@ -7,70 +7,153 @@
...
 
 # Localization
 _ = get_translation('spyder')
 
 
 # Actions
-class MainMenuActions:
+class ApplicationActions:
+    # Documentation related
+    SpyderDocumentationAction = "spyder documentation"
```
Comment (ccordoba12) [inline spyder/plugins/application/container.py:45]: Let's change this to `spyder_documentation_action` for consistency.

File: spyder/plugins/application/container.py [hunk 2]
```diff
@@ -7,70 +7,153 @@
...
+    from spyder.utils.environ import WinUserEnvDialog
 
 # Localization
 _ = get_translation('spyder')
 
 
 # Actions
-class MainMenuActions:
+class ApplicationActions:
+    # Documentation related
```
Comment (ccordoba12) [inline spyder/plugins/application/container.py:44]: Please change this comment to `# Help` to be in the same line of the comments below (i.e. to what menu these actions are going to be added).

File: spyder/plugins/application/plugin.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        # plugins). For other actions see ./container.py
+        self.restart_action = self.create_action(
+            ApplicationActions.SpyderRestart,
+            _("&Restart"),
+            icon=self.create_icon('restart'),
+            tip=_("Restart"),
+            triggered=self.restart,
+            context=Qt.ApplicationShortcut,
+            shortcut_context="_",
+            register_shortcut=True)
```
Comment (ccordoba12) [inline spyder/plugins/application/plugin.py:90]: Please move this action to the container and make it emit a signal (`sig_restart_requested`) when triggered. Then connect that signal to the `restart` method here.

File: spyder/plugins/console/widgets/main_widget.py [hunk 4]
```diff
@@ -51,7 +51,9 @@ class ConsoleWidgetActions:
     Environment = 'environment_action'
     ExternalEditor = 'external_editor_action'
     MaxLineCount = 'max_line_count_action'
-    Quit = 'quit_action'
+    # The name of the action needs to match name of the shortcut
```
Comment (ccordoba12) [inline spyder/plugins/console/widgets/main_widget.py:54]: ... needs to match **the** name of the ...

File: spyder/plugins/editor/plugin.py [hunk 5]
```diff
@@ -954,29 +954,70 @@ def get_plugin_actions(self):
...
+            section=FileMenuSections.New,
+            before_section=FileMenuSections.Restart)
+        # Open section
+        open_actions = [
             self.open_action,
             self.open_last_closed_action,
             self.recent_file_menu,
-            MENU_SEPARATOR,
-            MENU_SEPARATOR,
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:970]: Please unindent this bracket four spaces so it's at the same level of `open_actions`.

File: spyder/plugins/editor/plugin.py [hunk 6]
```diff
@@ -954,29 +954,70 @@ def get_plugin_actions(self):
...
+                before_section=FileMenuSections.Restart)
+        # Save section
+        save_actions = [
             self.save_action,
             self.save_all_action,
             save_as_action,
             save_copy_as_action,
             self.revert_action,
-            MENU_SEPARATOR,
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:984]: Unident 4 spaces.

File: spyder/plugins/editor/plugin.py [hunk 7]
```diff
@@ -954,29 +954,70 @@ def get_plugin_actions(self):
...
+                save_action,
+                menu_id=ApplicationMenus.File,
+                section=FileMenuSections.Save,
+                before_section=FileMenuSections.Restart)
+        # Print
+        print_actions = [
             print_preview_action,
             self.print_action,
-            MENU_SEPARATOR,
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:995]: Unident 4 spaces.

File: spyder/plugins/editor/plugin.py [hunk 8]
```diff
@@ -954,29 +954,70 @@ def get_plugin_actions(self):
...
+                section=FileMenuSections.Print,
+                before_section=FileMenuSections.Restart)
+        # Close
+        close_actions = [
             self.close_action,
-            self.close_all_action,
-            MENU_SEPARATOR,
-        ]
+            self.close_all_action
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:1006]: Unident 4 spaces.

File: spyder/plugins/editor/plugin.py [hunk 9]
```diff
@@ -954,29 +954,70 @@ def get_plugin_actions(self):
...
+                menu_id=ApplicationMenus.File,
+                section=FileMenuSections.Close,
+                before_section=FileMenuSections.Restart)
+        # Navigation
+        if sys.platform == 'darwin':
+            self.main.mainmenu.add_item_to_application_menu(
+                self.tab_navigation_actions,
+                menu_id=ApplicationMenus.File,
+                section=FileMenuSections.Navigation,
+                before_section=FileMenuSections.Restart)
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:1019]: Why did you add this? I mean, it seems it was not necessary before.

File: spyder/plugins/preferences/plugin.py [hunk 10]
```diff
@@ -25,8 +25,6 @@
 from spyder.config.base import _
 from spyder.config.main import CONF_VERSION
 from spyder.config.user import NoDefault
-from spyder.plugins.preferences.api import PreferencePages, GeneralConfigPage
```
Comment (ccordoba12) [inline spyder/plugins/preferences/plugin.py:28]: If `GeneralConfigPage` is not necessary anymore, please remove it to not leave dead code around.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @dalthviz for your work on this! This looks almost ready, except for the minor comments I left below.

File: spyder/plugins/application/container.py (hunk 1 above)
Comment (dalthviz) [reply spyder/plugins/application/container.py:45]: I think I left the names like that to be able to get the previously defined shortcut. If the name here changes the value defined for the action shortcut will not be picked. Maybe I need to do something else to take the shortcut then?

File: spyder/app/mainwindow.py [hunk 11]
```diff
@@ -3169,72 +3162,12 @@ def reset_spyder(self):
                "Do you want to continue?"),
              QMessageBox.Yes | QMessageBox.No)
         if answer == QMessageBox.Yes:
-            self.restart(reset=True)
+            self.application.restart(reset=True)
 
     @Slot()
     def restart(self, reset=False):
```
[the remark this answers is not in the document]
Comment (dalthviz) [reply spyder/app/mainwindow.py:3168]: I think I did the change but later I realized that the `restart` method is needed when registering plugins. Should I better revert calling `self.application.restart` to the previous `self.restart`?

File: spyder/plugins/editor/plugin.py (hunk 9 above)
Comment (dalthviz) [reply spyder/plugins/editor/plugin.py:1019]: That code was in the `mainwindow` (it was doing `self.editor.tab_navigation_actions`) so I added it here since I was already touching this code

File: spyder/plugins/preferences/plugin.py (hunk 10 above)
Comment (dalthviz) [reply spyder/plugins/preferences/plugin.py:28]: It is being used by the LSP

File: spyder/app/mainwindow.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/app/mainwindow.py:3129]: Yeah, I think so, so the presence of this method makes sense here.

File: spyder/plugins/application/container.py (hunk 1 above)
Comment (ccordoba12) [reply spyder/plugins/application/container.py:45]: Please leave a comment above then for now (like you did for other actions with associated shortctus).

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @dalthviz!
